ui/src/saros/ui/browser/SWTBrowserWrapper.java

```diff
@@ -0,0 +1,192 @@
+
+    browser.addDisposeListener(
+        (DisposeEvent e) -> {
+          for (Runnable runnable : runOnDisposalList) {
+            runnable.run();
+          }
+        });
+  }
+
+  @Override
+  public boolean setFocus() {
+    return this.syncExec(
+        () -> {
+          return browser.setFocus();
+        });
+  }
+
+  @Override
+  public void setSize(final int width, final int height) {
+    this.syncExec(() -> browser.setSize(width, height));
+  }
+
+  @Override
+  public String getUrl() {
+    return this.syncExec(
+        () -> {
+          return browser.getUrl();
+        });
+  }
+
+  @Override
+  public boolean loadUrl(final String url, final long timeout) {
+    browser.getShell().open();
+
+    this.state = ReadyState.LOADING;
+    browser.addProgressListener(readyStateListener);
+
+    browser.setUrl(url);
+    boolean successful = this.waitForPageLoaded(url, timeout);
+
+    browser.removeProgressListener(readyStateListener);
+
+    return successful;
+  }
+
+  /**
+   * Waits for a page to be loaded.
+   *
+   * @param url
+   * @param timeout
+   * @return
+   */
+  private boolean waitForPageLoaded(final String url, final long timeout) {
+
+    Runnable waitForLoaded =
+        () -> {
+          while (this.state != ReadyState.COMPLETE) {
+            if (!browser.getDisplay().readAndDispatch()) {
+              browser.getDisplay().sleep();
+            }
+          }
+        };
+
+    Thread t = new Thread(waitForLoaded);
+    t.setDaemon(true);
+    t.start();
+
+    try {
+      t.join(timeout);
+    } catch (InterruptedException e) {
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public void loadHtml(final String html) {
+    this.syncExec(() -> browser.setText(html));
+  }
+
+  @Override
+  public Object evaluate(final String jsCode) {
+    return this.syncExec(
+        () -> {
+          Object rv = browser.evaluate(jsCode);
+          System.out.println(rv);
+          return rv;
+        });
+  }
+
+  @Override
+  public boolean execute(String jsCode) {
+    return this.syncExec(
+        () -> {
+          return browser.execute(jsCode);
+        });
+  }
+
+  @Override
+  public void addBrowserFunction(final JavascriptFunction function) {
+    this.syncExec(
+        () -> {
+          BrowserFunction swtFunction =
+              new BrowserFunction(browser, function.getName()) {
+                @Override
+                public Object function(Object[] arguments) {
+                  return function.function(arguments);
+                }
+              };
+          return swtFunction;
+        });
+  }
+
+  @Override
+  public void runOnDisposal(Runnable runnable) {
+    runOnDisposalList.add(runnable);
+  }
+
+  private <V> V syncExec(final NoCheckedExceptionCallable<V> callable) {
+    final AtomicReference<V> r = new AtomicReference<V>();
+    final AtomicReference<RuntimeException> exception = new AtomicReference<RuntimeException>();
+    Display.getDefault()
```

Comment:
Display.getDefault() should not be called in Eclipse. You can "revive" / create a new display during the shutdown of the Eclipse IDE because the display is already disposed when the plugin tears down.

Reply:
I only moved this logic from the swt-browser-improved and forgot to adjust this call. `browser.getDisplay()` should be right.